### PR TITLE
ENH: .equals for Extension Arrays

### LINKED ISSUE #27081: API: .equals for Extension Arrays
We don't have a good method of testing for equality between EA's

```
In [19]: from pandas.tests.extension.decimal.array import DecimalArray                                                                                                                                                                                           

In [20]: from decimal import Decimal                                                                                                                                                                                                                             

In [21]: x = DecimalArray([Decimal('1'),Decimal('Nan')])                                                                                                                                                                                                         

In [22]: x == x                                                                                                                                                                                                                                                  
Out[22]: array([ True, False])

In [23]: x = pd.Series([1,np.nan], dtype='Int64').array                                                                                                                                                                                                          

In [24]: x == x                                                                                                                                                                                                                                                  
Out[24]: array([ True, False])

```

These *happen* to work with Series now because the null checks are handled at a higher level.
```
In [26]: x.equals(x)                                                                                                                                                                                                                                             
Out[26]: True

In [27]: x = pd.Series(DecimalArray([Decimal('1'),Decimal('Nan')]))                                                                                                                                                                                              

In [28]: x.equals(x)                                                                                                                                                                                                                                             
Out[28]: True
```
we could provide a default implementation that should just work in the presence of NA.

```
def equals(self, other):
    return ((self == other) | (self.isna() == other.isna())).all()
```

actually we should *also* implement a default ``__eq__``

### PR DESCRIPTION
**Pull request by dwhu**
- [x] closes #27081
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

--- Jan 3 10:00 PST

Hey team,

Looking for a bit of feedback on this API design and appropriate testing for this PR before I get too deep into it. Could you provide some guidance regarding how to appropriately handle updating this test? https://github.com/pandas-dev/pandas/blob/6f03e76f9d47ecfcfdd44641de6df1fc7dd57a01/pandas/tests/extension/base/ops.py#L162 It appears to contradict the issue #27081.

Appreciate the help,
Dave

### CODE REVIEW

**Comment by charlesdong1991** [comment]
thanks for your contribution! @dwhu 

in general here, if it is `NotImplemted`, and you implement some new functionalities, you could remove this one, and replace it with the corresponding tests to test your new implementations!

Good luck! and hope this helps

**Comment by dwhu** [comment]
> in general here, if it is `NotImplemted`, and you implement some new functionalities, you could remove this one, and replace it with the corresponding tests to test your new implementations!

@charlesdong1991 - Sounds good. Wanted to make sure I wasn't missing something or breaking a behavior contract by adding this behavior.

File: pandas/core/arrays/base.py [hunk 1]
```diff
@@ -350,6 +351,38 @@ def __iter__(self):
         for i in range(len(self)):
             yield self[i]
 
+    def __eq__(self, other: ABCExtensionArray) -> bool:
```
Comment (TomAugspurger) [inline pandas/core/arrays/base.py]: These will need to be added to the list of abstract methods in the "Notes" section of the class docstring.

File: pandas/core/arrays/base.py [hunk 2]
```diff
@@ -350,6 +351,38 @@ def __iter__(self):
...
+        ----------
+        other: ExtensionArray
+            The array to compare to this array.
+
+        Returns
+        -------
+        bool
+        """
+
+        raise AbstractMethodError(self)
```
Comment (TomAugspurger) [inline pandas/core/arrays/base.py]: Can this be `return ~(self == other)`?

File: pandas/core/arrays/base.py [hunk 3]
```diff
@@ -657,6 +690,23 @@ def searchsorted(self, value, side="left", sorter=None):
...
+        other: ExtensionArray
+            Array to compare to this Array.
+
+        Returns
+        -------
+        boolean
+            Whether the arrays are equivalent.
+
+        """
+        return ((self == other) | (self.isna() == other.isna())).all()
```
Comment (TomAugspurger) [inline pandas/core/arrays/base.py]: I think this also needs to ensure that the type of `other` matches the type of `self`. We don't want `pd.array([1, 2]).equals(np.array([1, 2]))` to be True (I think).

File: pandas/core/arrays/boolean.py [hunk 4]
```diff
@@ -314,6 +314,20 @@ def __getitem__(self, item):
             return self._data[item]
         return type(self)(self._data[item], self._mask[item])
 
+    def __eq__(self, other):
+        if not isinstance(other, BooleanArray):
+            return NotImplemented
+        return (
+            hasattr(other, "_data")
```
Comment (TomAugspurger) [inline pandas/core/arrays/boolean.py]: BooleanArray should already have a `__eq__` method (see `create_comparison_method` Was this needed?

**Review by TomAugspurger — COMMENTED** [review]
I think most of the `__eq__` implementations you've added can be removed.

@jschendel is in the middle of a PR fixing `IntervalArray.__eq__`, so ignore that for now if it's causing failures.

We'll need some base tests in `pandas/tests/extensions/base/ops.py` ensuring that `__eq__` and `__ne__` are implemented correctly. We'll want a few tests in `tests/extensions/base/methods.py` ensuring that `.equals` is tested. These tests are a bit tricky to write, so let us know if you have issues.

File: pandas/core/arrays/base.py (hunk 2 above)
Comment (dwhu) [reply pandas/core/arrays/base.py]: Done.

File: pandas/core/arrays/base.py (hunk 3 above)
Comment (dwhu) [reply pandas/core/arrays/base.py]: Done.

File: pandas/core/arrays/boolean.py (hunk 4 above)
Comment (dwhu) [reply pandas/core/arrays/boolean.py]: Not needed. Missed that in my WIP PR.

File: pandas/core/arrays/base.py (hunk 1 above)
Comment (dwhu) [reply pandas/core/arrays/base.py]: Done

File: pandas/core/arrays/base.py [hunk 5]
```diff
@@ -657,6 +691,25 @@ def searchsorted(self, value, side="left", sorter=None):
...
+        other: ExtensionArray
+            Array to compare to this Array.
+
+        Returns
+        -------
+        boolean
+            Whether the arrays are equivalent.
+
+        """
+        return isinstance(other, self.__class__) and (
```
Comment (TomAugspurger) [inline pandas/core/arrays/base.py]: Perhaps `type(self) == type(other)`? It's not clear to me whether a subclass should compare `.equals` to a parent class. Thoughts @jreback?

Comment (TomAugspurger) [reply pandas/core/arrays/base.py]: Will want some length checking here too.

File: pandas/core/arrays/integer.py [hunk 6]
```diff
@@ -376,6 +376,15 @@ def __getitem__(self, item):
 
         return type(self)(self._data[item], self._mask[item])
 
+    def __eq__(self, other):
+        return (
+            isinstance(other, IntegerArray)
+            and hasattr(other, "_data")
```
Comment (TomAugspurger) [inline pandas/core/arrays/integer.py]: We already have `IntegerArray.__eq__` on master. This shouldn't be necessary.

File: pandas/tests/extension/base/methods.py [hunk 7]
```diff
@@ -358,3 +358,18 @@ def test_repeat_raises(self, data, repeats, kwargs, error, msg, use_numpy):
                 np.repeat(data, repeats, **kwargs)
             else:
                 data.repeat(repeats, **kwargs)
+
+    def test_equals(self, data, na_value):
+        cls = type(data)
+        ser = pd.Series(cls._from_sequence(data, dtype=data.dtype))
+        na_ser = pd.Series(cls._from_sequence([na_value], dtype=data.dtype))
+
+        assert data.equals(data)
```
Comment (TomAugspurger) [inline pandas/tests/extension/base/methods.py]: make all these `assert <condition> is True`, to ensure that `True` or `False` is actually returned

File: pandas/tests/extension/base/methods.py [hunk 8]
```diff
@@ -358,3 +358,18 @@ def test_repeat_raises(self, data, repeats, kwargs, error, msg, use_numpy):
...
+        assert ser.equals(ser)
+        assert na_ser.equals(na_ser)
+
+        assert not data.equals(na_value)
+        assert not na_ser.equals(ser)
+        assert not ser.equals(na_ser)
+        assert not ser.equals(0)
+        assert not na_ser.equals(0)
```
Comment (TomAugspurger) [inline pandas/tests/extension/base/methods.py]: Add test for other inputs like an ndarray.

Add a test for unequal length things.

File: pandas/tests/extension/base/ops.py [hunk 9]
```diff
@@ -132,10 +132,8 @@ class BaseComparisonOpsTests(BaseOpsUtil):
     def _compare_other(self, s, data, op_name, other):
         op = self.get_op_from_name(op_name)
         if op_name == "__eq__":
-            assert getattr(data, op_name)(other) is NotImplemented
```
Comment (TomAugspurger) [inline pandas/tests/extension/base/ops.py]: We definitely don't want to remove this :) You'll need to revert it and whatever change required it :)

File: pandas/tests/extension/base/ops.py [hunk 10]
```diff
@@ -158,13 +156,3 @@ def test_compare_array(self, data, all_compare_operators):
         s = pd.Series(data)
         other = pd.Series([data[0]] * len(data))
         self._compare_other(s, data, op_name, other)
-
-    def test_direct_arith_with_series_returns_not_implemented(self, data):
```
Comment (TomAugspurger) [inline pandas/tests/extension/base/ops.py]: Undo this change.

File: pandas/tests/extension/base/methods.py (hunk 8 above)
Comment (dwhu) [reply pandas/tests/extension/base/methods.py]: Done.

File: pandas/tests/extension/base/methods.py (hunk 7 above)
Comment (dwhu) [reply pandas/tests/extension/base/methods.py]: I ran into a funny issue doing this: https://github.com/numpy/numpy/issues/5942
I'm going to keep this as is now to keep the test consistent within itself.

`.equals()` returns a `np.bool_` type. As result when you compare it to bool type `True` using `is` (i.e. `some_np_bool_True_val is True == False` 

Coercing `np.bool_` to `bool` using the `bool()` constructor did not work either. `== True` causes PEP8 errors asking me to use `assert <condition>` or `assert <condition> is True`.

Open to suggestions if you have any.

File: pandas/core/arrays/base.py (hunk 5 above)
Comment (dwhu) [reply pandas/core/arrays/base.py]: Added `type(self) == type(other)` and length checks.

**Comment by WillAyd** [comment]
@dwhu looks like CI is red - can you take a look and fix up the merge conflict?

**Comment by WillAyd** [comment]
Closing as I think this has gone stale but ping if you'd like to pick back up. Thanks!

**Comment by jorisvandenbossche** [comment]
@dwhu I am going to look into further updating this, if that's OK with you

File: pandas/tests/extension/base/ops.py (hunk 10 above)
Comment (jorisvandenbossche) [reply pandas/tests/extension/base/ops.py:169]: > Same as the other comment. The whole PR is intended to break this test since we implement __eq__ and __ne__.

The reason this needs to be kept is that, although `__eq__` is a required method on an ExtensionArray now, we still require it to return NotImplemented when `other` is a Series or Index.

File: pandas/tests/extension/base/ops.py (hunk 9 above)
Comment (jorisvandenbossche) [reply pandas/tests/extension/base/ops.py:142]: @TomAugspurger I think this is actually correct to change this. As far as I understand, we only tested this as the base version (if you don't define an `__eq__` on your class, python will return this, I assume). But now we changed the base class to actually have a `__eq__` but that raise an AbstractError, so it needs to always be implemented, and thus can return something here that is not NotImplemented.

Comment (TomAugspurger) [reply pandas/tests/extension/base/ops.py:142]: What's `other` here? If it's a Series / DataFrame we need to make sure to return `NotImplemented` right? I thought that was the intent of this test, though I might misunderstand.

Comment (jorisvandenbossche) [reply pandas/tests/extension/base/ops.py:142]: `data` is the EA, and other can eg be a scalar. 

There is a separate test just below (that I added back) that explicitly tests that ``NotImplemented`` is returned for `EA.__eq__(Series/Index)`

**Comment by jorisvandenbossche** [comment]
OK, I updated this with the following things:

- Made `__eq__` into an abstract method so it is required to implement (another option could be to `return NotImplemented` as default, but that's actually wrong, as it results in not doing element-wise comparison for `EA == EA`, but just returning False)
- Fixed the `equals` imlementation: it became a bit longer to ensure we check for BooleanArray (if the `==` comparison returns a nullable boolean array, we need to fill NAs with False)
- Added `ExtensionBlock.equals` (overriding the base Block method) to ensure we dispatch to the new EA method
- A bit of clean-up of the tests

Should be ready for a new review now @TomAugspurger @jbrockmendel

File: pandas/core/arrays/base.py [hunk 11]
```diff
@@ -333,6 +335,18 @@ def __iter__(self):
         for i in range(len(self)):
             yield self[i]
 
+    def __eq__(self, other: Any) -> Union[np.ndarray, "ExtensionArray"]:
+        """
+        Return for `self == other` (element-wise equality).
```
Comment (TomAugspurger) [inline pandas/core/arrays/base.py:340]: I think we should have some guidance here that if `other` is a Series then you should return NotImplemented.

File: pandas/core/arrays/base.py [hunk 12]
```diff
@@ -682,6 +696,33 @@ def searchsorted(self, value, side="left", sorter=None):
...
+            Array to compare to this Array.
+
+        Returns
+        -------
+        boolean
+            Whether the arrays are equivalent.
+
+        """
+        if not type(self) == type(other):
+            return False
```
Comment (TomAugspurger) [inline pandas/core/arrays/base.py:715]: We should verify that this is the behavior we want. Namely

1. other array-likes are not equivalent, even if they are all equal.
2. *subclasses* are not equivalent, even if they are all equal.

The first seems fine. Not sure about the second.

File: pandas/tests/extension/base/methods.py [hunk 13]
```diff
@@ -421,3 +421,29 @@ def test_repeat_raises(self, data, repeats, kwargs, error, msg, use_numpy):
                 np.repeat(data, repeats, **kwargs)
             else:
                 data.repeat(repeats, **kwargs)
+
+    def test_equals(self, data, na_value, as_series):
```
Comment (TomAugspurger) [inline pandas/tests/extension/base/methods.py:425]: Can you also include `DataFrame` in this parameterization, or do it somewhere else? Just to ensure coverage of the changes to `blocks.py`.

**Review by TomAugspurger — COMMENTED** [review]
Looks nice. Can you add a release note?

File: pandas/core/arrays/base.py (hunk 12 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/base.py:715]: I was planning to open an issue about this after this PR (and keep it strict here), because this is right now a bit inconsistent within pandas, and might require a more general discussion / clean-up (eg Series.equals is more strict (requires same dtype) than Index.equals ...) 
But we can certainly also have the discussion here.

File: pandas/core/arrays/base.py [hunk 14]
```diff
@@ -333,6 +335,18 @@ def __iter__(self):
         for i in range(len(self)):
             yield self[i]
 
+    def __eq__(self, other: Any) -> Union[np.ndarray, "ExtensionArray"]:
+        """
+        Return for `self == other` (element-wise equality).
+        """
+        raise AbstractMethodError(self)
+
+    def __ne__(self, other: Any) -> Union[np.ndarray, "ExtensionArray"]:
```
Comment (jschendel) [inline pandas/core/arrays/base.py:344]: This should allow us to remove the equivalent `IntervalArray` definition here:

https://github.com/pandas-dev/pandas/blob/d149f4109de8610ccc4e3d229c037047b728a49e/pandas/core/arrays/interval.py#L609-L610

File: pandas/core/arrays/base.py (hunk 12 above)
Comment (TomAugspurger) [reply pandas/core/arrays/base.py:715]: Happy to be strict for now.

File: pandas/core/arrays/base.py (hunk 11 above)
Comment (jbrockmendel) [reply pandas/core/arrays/base.py:340]: could use ops.common.unpack_zerodim_and_defer

File: pandas/core/arrays/base.py (hunk 14 above)
Comment (jbrockmendel) [reply pandas/core/arrays/base.py:344]: `Union[np.ndarray, "ExtensionArray"]` -> `ArrayLike`

File: pandas/core/arrays/base.py [hunk 15]
```diff
@@ -682,6 +696,33 @@ def searchsorted(self, value, side="left", sorter=None):
...
+            return False
+        elif not len(self) == len(other):
+            return False
+        else:
+            equal_values = self == other
+            if isinstance(equal_values, ExtensionArray):
+                # boolean array with NA -> fill with False
+                equal_values = equal_values.fillna(False)
+            equal_na = self.isna() & other.isna()
+            return (equal_values | equal_na).all().item()
```
Comment (jbrockmendel) [inline pandas/core/arrays/base.py:724]: is the `.item()` necessary?

File: pandas/core/arrays/base.py (hunk 11 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/base.py:340]: > could use ops.common.unpack_zerodim_and_defer

This method is to be implemented by EA authors, so those can't use that helper (unless we expose somewhere a public version of this). 

(we could of course use that for our own EAs, but this PR is not changing any existing `__eq__` implementation at the moment)

Comment (jorisvandenbossche) [reply pandas/core/arrays/base.py:340]: > I think we should have some guidance here that if other is a Series then you should return NotImplemented.

Added a comment about that

File: pandas/core/arrays/base.py (hunk 12 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/base.py:715]: I made it even more strict (same dtype, not just same class), and added a test for that. 
Will open an issue for the general discussion.

File: pandas/core/arrays/base.py (hunk 15 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/base.py:724]: > is the .item() necessary?

It's to have a python bool, instead of a numpy bool, as result. 
I added a comment to the tests to make it explicit this is the reason we are asserting with `is True/False` (and later on we don't inadvertedly "clean" that up)

**Comment by jorisvandenbossche** [comment]
@simonjayhawkins how do I solve this typing error?

```
pandas/core/arrays/base.py:338: error: Return type "Union[Any, ExtensionArray]" of "__eq__" incompatible with return type "bool" in supertype "object"
```

So our annotation of `__eq__` is incompatible with the base "object" class, but we do this on purpose. So is there another way to solve this than to suppress the error in some way?

File: pandas/core/arrays/base.py (hunk 14 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/base.py]: @jschendel done!

File: pandas/core/arrays/base.py (hunk 12 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/base.py:721]: I opened an issue for this at https://github.com/pandas-dev/pandas/issues/33940

In the end, it seems mainly to come to whether the dtype should exactly be equal or not. 

Since for EAs, the dtype is right now tied to the array class, using equal dtype for now also implies the same class (no additional check to allow sublcasses).

File: pandas/core/arrays/base.py (hunk 15 above)
Comment (jbrockmendel) [reply pandas/core/arrays/base.py:732]: thanks for clarifying.  i was recently reminded that `np.array(np.timedelta64(1234, "ns")).item()` gives an int instead of timedelta64, so im now cautious around `.item()`

**Comment by jorisvandenbossche** [comment]
Any more comments on this PR?

**Comment by jreback** [comment]
will look soon

File: pandas/core/arrays/base.py [hunk 16]
```diff
@@ -335,7 +335,7 @@ def __iter__(self):
         for i in range(len(self)):
             yield self[i]
 
-    def __eq__(self, other: Any) -> ArrayLike:
+    def __eq__(self, other: Any) -> ArrayLike:  # type: ignore[override] # NOQA
```
Comment (jbrockmendel) [inline pandas/core/arrays/base.py:338]: im not familiar with the `ignore[overrride] # NOQA` pattern.  do we use that elsewhere?

Comment (jorisvandenbossche) [reply pandas/core/arrays/base.py:338]: I thought I posted the mypy issue, but apparently forgot.

So the problem here is that mypy expects a "bool" return for `__eq__` since the base `object` is typed that way in the stubs. In https://github.com/python/mypy/issues/2783, the recommended way to solve this is the above with `# type: ignore[override]`. 

But flake8 doesn't like that, hence also the #NOQA (https://github.com/PyCQA/pyflakes/issues/475). 

This was already done in another PR as well:

https://github.com/pandas-dev/pandas/blob/7aa710a4d4d3b8515a723a366e50f49b327adc23/pandas/core/frame.py#L5062

**Comment by jorisvandenbossche** [comment]
@jreback you wanted to take a look here?

File: doc/source/whatsnew/v1.1.0.rst [hunk 17]
```diff
@@ -150,6 +150,7 @@ Other enhancements
   such as ``dict`` and ``list``, mirroring the behavior of :meth:`DataFrame.update` (:issue:`33215`)
 - :meth:`~pandas.core.groupby.GroupBy.transform` and :meth:`~pandas.core.groupby.GroupBy.aggregate` has gained ``engine`` and ``engine_kwargs`` arguments that supports executing functions with ``Numba`` (:issue:`32854`, :issue:`33388`)
 - :meth:`~pandas.core.resample.Resampler.interpolate` now supports SciPy interpolation method :class:`scipy.interpolate.CubicSpline` as method ``cubicspline`` (:issue:`33670`)
+- The ``ExtensionArray`` class has now an ``equals`` method, similarly to ``Series.equals()`` (:issue:`27081`).
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst:153]: you can add the doc reference here

File: pandas/core/internals/blocks.py [hunk 18]
```diff
@@ -1864,6 +1864,11 @@ def where(
 
         return [self.make_block_same_class(result, placement=self.mgr_locs)]
 
+    def equals(self, other) -> bool:
+        if self.dtype != other.dtype or self.shape != other.shape:
```
Comment (jreback) [inline pandas/core/internals/blocks.py:1868]: why do you need this check here? (its already done on the .values no?)

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. a couple of minor comment.s

File: pandas/core/internals/blocks.py (hunk 18 above)
Comment (jorisvandenbossche) [reply pandas/core/internals/blocks.py:1868]: Hmm, good question. I just copied that from the base Block.equals method (so it's done there as well). There, `array_equivalent` is used. Maybe that is less strict and the extra check is needed, which is not needed here.

Comment (jorisvandenbossche) [reply pandas/core/internals/blocks.py:1868]: At least from the docstring of `array_equivalent` (not sure how up to date that is), using that method indeed requires a more strict check in advance of calling it:

https://github.com/pandas-dev/pandas/blob/3ed7dff48bb4e8c7c0129283ff51eccea3a0f861/pandas/core/dtypes/missing.py#L360-L363

But so, for EA.equals that is not needed, so will remove that check here.

**Comment by jorisvandenbossche** [comment]
Travis failures are unrelated (network issues, S3)
